theano/gpuarray/opt.py

```diff
@@ -319,7 +319,8 @@ def apply(self, fgraph):
 
             context_name = None
             for i in [mapping[i] for i in node.inputs]:
-                if isinstance(i.type, GpuArrayType):
+                target = getattr(i.tag, 'target', None)
+                if (isinstance(i.type, GpuArrayType) and target != 'cpu'):
```

Comment:
What about handling for the inputs ? You are handling the condition as `!='cpu'` in line:288 and a similar condition check on the 'cpu' string in line:233(gpuarray/opt.py). What about those checks? Shall I remove them as well ? 
And what I understand from what @abergeron says is, we need to do something like line:83 of gpuarray/basic_ops.py ?